### PR TITLE
Ready Integration Testing for Prime Time

### PR DESCRIPTION
**Pull request by jmchilton**
Integration tests were added in https://github.com/galaxyproject/galaxy/pull/2078 and a74cda60f551e8cbc8a18d53d2f788f7ea2ea73f - and expanded by @mvdbeek for Conda testing in b5fad4a04804e6cefec335fe888bf616c7eb7e0d.

From a74cda60f551e8cbc8a18d53d2f788f7ea2ea73f, "IntegrationTestCases differ from functional tests in that they start up a Galaxy instance configured on a per test class basis - so this new class of "integration tests" can test configuration modalities of Galaxy that cannot be tested with traditional Galaxy functional tests."

This PR adds the ability to run integration tests via ``run_tests.sh`` and fixes up the Jenkins script I sketched out previously but didn't touch for executing them. It also refactors how integration tests can leverage utilities from the API test framework.

This PR adds a test case for various advanced job runner functionality - job resource parameters and re-submission of jobs based on hitting walltime and memory limits. These were previously not tested at all. In order to implement this testing I also enhanced the local job runner to leverage the job state handling framework implemented by @natefoo and previously only available to the asynchronous job runners.

xref #3245

### CODE REVIEW

**Comment by natefoo** [comment]
Hooray for not having to duplicate the resource param parsing in every rule. Doing this probably should have been obvious but I'm never one for getting the obvious...

File: lib/galaxy/jobs/__init__.py [hunk 1]
```diff
@@ -764,7 +764,30 @@ def convert_legacy_destinations(self, job_runners):
...
+
+        if job is None:
+            job = self.get_job()
+
+        app = self.app
+        param_values = job.get_param_values( app, ignore_errors=True )
+        resource_params = {}
+        try:
+            resource_params_raw = param_values[ "__job_resource" ]
+            if resource_params_raw[ "__job_resource__select" ].lower() in [ "1", "yes", "true" ]:
```
Comment (bgruening) [inline lib/galaxy/jobs/__init__.py:782]: Maybe a modified one of this function? https://github.com/galaxyproject/galaxy/blob/71cea6604d43c5fe6215f5656462ba6c1af69bb6/lib/galaxy/web/framework/helpers/__init__.py#L109

File: lib/galaxy/jobs/__init__.py [hunk 2]
```diff
@@ -764,7 +764,30 @@ def convert_legacy_destinations(self, job_runners):
...
+                for key, value in resource_params_raw.iteritems():
```
Comment (bgruening) [inline lib/galaxy/jobs/__init__.py]: This is not python3, but I guess it does not count here?

**Review by bgruening — APPROVED** [review]

File: lib/galaxy/jobs/__init__.py (hunk 2 above)
Comment (nsoranzo) [reply lib/galaxy/jobs/__init__.py]: We should use `items()` everywhere unless there is a performance concern, in which case there is `six.iteritems()` .

Comment (jmchilton) [reply lib/galaxy/jobs/__init__.py]: We should use items - I'll update this.

File: lib/galaxy/jobs/__init__.py (hunk 1 above)
Comment (jmchilton) [reply lib/galaxy/jobs/__init__.py:782]: I don't want a dependency between galaxy.web.framework.helpers and galaxy.jobs. I guess we don't want to expand the web stuff to allow ``1`` as well - we should be moving toward JSON booleans in the API - not strings for web stuff. So I would say we keep it like it is for now? If you really want me to modify it - I can switch this out for a generic tool handling boolean thing I guess - but I'd prefer not to.

**Comment by jmchilton** [comment]
Okay - so I have significantly expanded the scope of this PR. It now covers essentially all of https://github.com/galaxyproject/galaxy/issues/3293 as well.
 - There is a new job runner failure type (in addition to memory and walltime) called "unknown_error" that can be a re-submission condition.
 - The re-submit condition can now be a simple Python expression (a whitelisted expression despite the fact we absolutely allow arbitrary code execution in job_conf.xml in other ways).
 - The re-submit condition includes new variables for ``attempts``, ``seconds_since_queued``, and ``seconds_running`` to allow various re-submission combinations (re-submit long jobs, short jobs, X number of times, and logical combinations thereof).
 - The re-submit condition can now include a delay (in either absolute seconds or some expression - for instance using ``attempt`` or ``seconds_since_queued`` to implement back-off behaviors). 
 - Integration tests for all of thee above.

File: lib/galaxy/jobs/__init__.py (hunk 1 above)
Comment (bgruening) [reply lib/galaxy/jobs/__init__.py:782]: Not needed at all.

**Comment by natefoo** [comment]
:100:

**Pull request merged**